docs/design/federation-phase-1.md

```diff
@@ -0,0 +1,372 @@
+   sufficient, workloads should be automatically distributed to other
+   clusters.
++ **Vendor lock-in avoidance:** Customers want to spread their
+   workloads on different cloud providers, and can easily increase or
+   decrease the workload proportion of a specific provider.
++ **Cluster Size Enhancement:** Currently K8S cluster can only support
+a limited size. While the community is actively improving it, it can
+be expected that cluster size will be a problem if K8S is used for
+large workloads or public PaaS infrastructure. While we can separate
+different tenants to different clusters, it would be good to have a
+unified view.
+
+Here are the functionality requirements derived from above use cases:
+
++  Clients can register and deregister clusters.
++ Workloads should be spread to different clusters according to the
+   workload distribution policy.
++ Pods running on different clusters are able to discover each other
+   and communicate with each other.
++ Traffic to these pods should be load balanced among clusters.
++ The control plane needs to know when a cluster is down, and migrate
+   the workloads to other clusters.
++ Clients have a unified view and a central control point for above
+   activities.
+
+## SCOPE
+
+It’s difficult to have a perfect design with one click that implements
+all the above requirements. Therefore we will go with an iterative
+approach to design and build the system. This document describes the
+phase one of the whole work.  In phase one we will cover only the
+following objectives:
+
++ Define the basic building blocks and API objects of control plane
++ Implement a basic end-to-end workflow
+   + Clients register federated clusters
+   + Clients submit a workload
+   + The workload is distributed to different clusters
+   + Service discovery
+   + Load balancing
+
+The following parts are NOT covered in phase one:
+
++ Authentication and authorization
++ Deployment units other than replication controller and service
++ Complex distribution policy of workloads
++ Service affinity and migration
+
+## ARCHITECTURE
+
+The overall architecture of a control plane is shown as following:
+
+![Ubernetes Architecture](ubernetes-design.png)
+
+Some design principles we are following in this architecture:
+
+1. Keep the underlying K8S clusters independent. They should have no
+   knowledge of control plane or of each other.
+1. Keep the Ubernetes API interface compatible with K8S API as much as
+   possible.
+1. Re-use concepts from K8S as much as possible. This reduces
+customers’ learning curve and is good for adoption.  Below is a brief
+description of each module contained in above diagram.
+
+## API Server
+
+The API Server in Ubernetes control plane works just like the API
+Server in K8S. It talks to the distributed storage to persist,
+retrieve and watch API objects.  We still use `etcd` as the distributed
+storage so customers don’t need to learn and manage a different
+storage system.
+
+## Scheduler
```

Comment:
1) This is missing some kind of high-level introductory sentence. Since the Kubernetes scheduler schedules pods to machines, you should open by saying "The Ubernetes scheduler schedules X to Y." I actually wonder if you should give it a different name than "scheduler." It's sort of a meta-controller, really -- it creates and manages controllers (or more correctly, the objects managed by controllers, but we don't really distinguish those concepts very clearly in Kubernetes).

2) What is a "pending replication controller"? I'm not saying there's anything wrong with the idea, but it's a completely new concept (there is no "pending" notion for replication controllers in regular Kubernetes) so it needs an explanation.

3) You should choose a different term than "sub-RC." The problem is that when you generalize it to "sub-resource" you conflict with another definition of "subresource" that we're already using in Kubernetes. My second problem with the term "sub-RC" is that it sounds like it's a different kind than the normal single-cluster ReplicationController, but IIUC from Quinton's doc, it's actually exactly the same object, just one that is managed by Ubernetes control plane rather than end-user.